README.md
# MMM-CaravanPiTemperature

This is a module for the [MagicMirror²](https://github.com/MichMich/MagicMirror/).

Todo: Insert description here!

## Using the module

To use this module, add the following configuration block to the modules array in the `config/config.js` file:
```js
var config = {
    modules: [
        {
            module: 'MMM-CaravanPiTemperature',
            config: {
                // See below for configurable options
            }
        }
    ]
}
```

## Configuration options

| Option           | Description
|----------------- |-----------
| `option1`        | *Required* DESCRIPTION HERE
| `option2`        | *Optional* DESCRIPTION HERE TOO <br><br>**Type:** `int`(milliseconds) <br>Default 60000 milliseconds (1 minute)


CHANGELOG.md
# MMM-CaravanPiTemperature Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).


## [1.0.0] - Unreleased

First public release


node_helper.js
/* ---------------------------------------------------------------------
 * Magic Mirror
 * Module: MMM-CaravanPiTemperature
 *
 * CaravanPi Module
 * see https://github.com/spitzlbergerj/CaravanPi for more Information 
 *		 about the DIY project 
 *
 * By Josef Spitzlberger	http://spitzlberger.de
 * MIT Licensed.
 */

const NodeHelper = require("node_helper")
var async = require('async');
var exec = require('child_process').exec;

//globale Variable, weil diese ansonsten in fillValueList unbekannt
valueListNHCaravanPiTemperature = [];

module.exports = NodeHelper.create({
	
	start: function() {
		//console.error('Starting node helper: ' + this.name);
	},
	
	socketNotificationReceived: function(notification, payload) {
		var self = this;
		console.error('node_helper: ' + notification);
		
		switch(notification) {
			case "CONFIG":
				this.config = payload.config;
				valueListNHCaravanPiTemperature = payload.valueList;
				// first call
				self.getValues(valueListNHCaravanPiTemperature);
				// interval call
				setInterval(function() {
					self.getValues(valueListNHCaravanPiTemperature);
				}, this.config.updateInterval);
				break
		}
	},
	
	getValues: function(valueList) {
		var self = this;
		var cmdPart = "tail -1 " + self.config.valueDir + "/";
		var cmd = "";
		var i = 0;
		
		while (i<valueList.length) {
			cmd = cmdPart + valueList[i]["file"]
			console.error('node_helper - cmd', cmd);
			exec(cmd,"",this.fillValueList);

			i+=1;
		}

		console.error('node_helper - getValues - valueList after', valueListNHCaravanPiTemperature[0], valueListNHCaravanPiTemperature[1]);
		self.sendSocketNotification('VALUES', valueListNHCaravanPiTemperature);
	},
	
	fillValueList: function (err, stdout, stderr) {
		var i = 0;
		
		if (err) {
			console.error('node_helper - fillValueList - Fehler:', err, stderr);
			return;
		}
		var resSplit = stdout.split(' ');
		var sensorID = resSplit[0];
		
		console.error('node_helper - fillValueList ', stdout, sensorID);
		
		while (i<valueListNHCaravanPiTemperature.length) {
			if (sensorID === valueListNHCaravanPiTemperature[i]["file"]) {
				valueListNHCaravanPiTemperature[i]["datetime"] = resSplit[1].substring(6,8)+"."+resSplit[1].substring(4,6)+"."+resSplit[1].substring(0,4)+" "+resSplit[1].substring(8,10)+":"+resSplit[1].substring(10,12);
				valueListNHCaravanPiTemperature[i]["temperature"] = resSplit[2];
			}
			i+=1;
		}
		return;
	},

})

MMM-CaravanPiClimate.js
/* ---------------------------------------------------------------------
 * Magic Mirror
 * Module: MMM-CaravanPiClimate
 *
 * CaravanPi Module
 * see https://github.com/spitzlbergerj/CaravanPi for more Information 
 *     about the DIY project 
 *
 * By Josef Spitzlberger http://spitzlberger.de
 * MIT Licensed.
 */

Module.register("MMM-CaravanPiClimate",{

defaults:{
	valueDir: "/home/pi/CaravanPi/values",
	updateInterval: 100000, // milliseconds
	tempUnit: " °C",
	humUnit: " %",
	pressUnit: " hPa",
	tempPrecision: 2,
	humPrecision: 2,
	pressPrecision: 2,
	showDate: true,
	sensors: [
		{
			name: "Innenraum",
			file: "BME280-96-118",
		},
		{
			name: "Außenbereich",
			file: "BME280-96-119",
		},
	],
	localeStr: 'de-DE',
	style: "lines",
},

valueList:[],

start: function (){
	Log.log('Starting module: ' + this.name);
	
	this.valueList = new Array();
	var i = 0;
	Log.log('sensors: ', this.config.sensors.length, this.config.sensors);
	while(i<this.config.sensors.length){
		this.valueList[i] = new Object();
		this.valueList[i]["name"] = this.config.sensors[i]["name"];
		this.valueList[i]["file"] = this.config.sensors[i]["file"];
		this.valueList[i]["datetime"] = this.translate('LOADING');
		this.valueList[i]["temperature"] = "0";
		this.valueList[i]["pressure"] = "0";
		this.valueList[i]["humidity"] = "0";
		i+=1;
	}
	Log.log('valueList: ', this.valueList);
	this.sendSocketNotification(
		'CONFIG',
		{
			config: this.config,
			valueList: this.valueList,
		});
},

// Get translations
getTranslations: function() {
	return {
		en: "translations/en.json",
		de: "translations/de.json"
	}
},

// Get the Module CSS
getStyles: function() {
	return ["MMM-CaravanPiClimate.css"];
},


getDom: function(){
	var table = document.createElement("table");
	table.border = 0;
	
	if (this.config.style == "boxes") {
		var boxRow = document.createElement("tr");
		boxRow.className = 'sensorBoxRow';
		boxRow.vAlign = 'top';
	}

	var i = 0;
	while (i<this.config.sensors.length) {
		var temperatureStr = this.prepareAttribute("TEMPERATURE", this.valueList[i]["temperature"], this.config.tempPrecision, this.config.tempUnit);
		var pressureStr = this.prepareAttribute("PRESSURE", this.valueList[i]["pressure"], this.config.pressPrecision, this.config.pressUnit);
		var humidityStr = this.prepareAttribute("HUMIDITY", this.valueList[i]["humidity"], this.config.humPrecision, this.config.humUnit);

		if (this.config.style == "lines") {
			var row = document.createElement("tr");
			row.className = 'sensorContainer';
			row.vAlign = 'top';
			
			var rowSensor = document.createElement("td");
			rowSensor.className = 'sensorName';
			rowSensor.width = '120px';
			rowSensor.appendChild(document.createTextNode(this.valueList[i]["name"]));
			
			var rowTemperature = document.createElement("td");
			rowTemperature.className = 'sensorTemp';
			rowTemperature.width = '60px';
			rowTemperature.appendChild(document.createTextNode(temperatureStr));
			
			var rowPressure = document.createElement("td");
			rowPressure.className = 'sensorPress';
			rowPressure.width = '60px';
			rowPressure.appendChild(document.createTextNode(pressureStr));
			
			var rowHumidity = document.createElement("td");
			rowHumidity.className = 'sensorHum';
			rowHumidity.width = '60px';
			rowHumidity.appendChild(document.createTextNode(humidityStr));
			
			var rowDate = document.createElement("td");
			rowDate.className = 'sensorDate';
			rowDate.width = '60px';
			rowDate.appendChild(document.createTextNode(this.valueList[i]["datetime"]));
			
			// Building of the table row
			row.appendChild(rowSensor);
			row.appendChild(rowTemperature);
			row.appendChild(rowPressure);
			row.appendChild(rowHumidity);
			
			if(this.config.showDate === true) {
				row.appendChild(rowDate);
			}
			
			table.appendChild(row);
		}
		else if (this.config.style == "boxes") {
			var boxRowElement = document.createElement("td");
			
			var tableInner = document.createElement("table");
			tableInner.border= '1px';
			tableInner.margin = '2px';
	
			var row1 = document.createElement("tr");
			row1.className = 'sensorContainer';
			row1.align = 'center';
			row1.vAlign = 'top';
			row1.border_bottom = "1px";
			row1.margin_bottom = "5px";
			
			var rowSensor = document.createElement("td");
			rowSensor.className = 'sensorName';
			rowSensor.appendChild(document.createTextNode(this.valueList[i]["name"]));
			
			row1.appendChild(rowSensor);
			
			var row2 = document.createElement("tr");
			row2.className = 'sensorContainer';
			row2.align = 'center';
			row2.vAlign = 'top';
			
			var rowTemperature = document.createElement("td");
			rowTemperature.className = 'sensorTemp';
			rowTemperature.appendChild(document.createTextNode(temperatureStr));
			
			row2.appendChild(rowTemperature);
			
			var row3 = document.createElement("tr");
			row3.className = 'sensorContainer';
			row3.align = 'center';
			row3.vAlign = 'top';
			
			var rowPressure = document.createElement("td");
			rowPressure.className = 'sensorPress';
			rowPressure.appendChild(document.createTextNode(pressureStr));
			
			row3.appendChild(rowPressure);
			
			var row4 = document.createElement("tr");
			row4.className = 'sensorContainer';
			row4.align = 'center';
			row4.vAlign = 'top';
			
			var rowHumidity = document.createElement("td");
			rowHumidity.className = 'sensorHum';
			rowHumidity.appendChild(document.createTextNode(humidityStr));
			
			row4.appendChild(rowHumidity);
			
			var row5 = document.createElement("tr");
			row5.className = 'sensorContainer';
			row5.align = 'center';
			row5.vAlign = 'top';
			
			var rowDate = document.createElement("td");
			rowDate.className = 'sensorDate';
			rowDate.appendChild(document.createTextNode(this.valueList[i]["datetime"]));
			
			row5.appendChild(rowDate);
			
			// Building of the table rows
			tableInner.appendChild(row1);
			tableInner.appendChild(row2);
			tableInner.appendChild(row3);
			tableInner.appendChild(row4);
			
			if(this.config.showDate === true) {
				tableInner.appendChild(row5);
			}
			boxRowElement.appendChild(tableInner)
			boxRow.appendChild(boxRowElement);
		}
		i+=1;
	}
	
	if (this.config.style == "boxes") {
		table.appendChild(boxRow);
	}

	var wrapper = document.createElement("div")
	wrapper.className = "MMM-CaravanPiClimate";
	
	wrapper.innerHTML = table.outerHTML;
	return wrapper
},

/*
notificationReceived: function(notification, payload, sender){
	switch(notification) {
		case "DOM_OBJECTS_CREATED":
			var timer = setInterval(()=>{
				var countElm = document.getElementById("COUNT")
				this.sendSocketNotification("DO_YOUR_JOB", this.count)
				this.count++
			}, 1000)
			break
	}
},

*/

socketNotificationReceived: function(notification, payload){
	Log.log('MMM-Systemvalues: socketNotificationReceived ' + notification + payload);
	switch(notification) {
		case "VALUES":
			this.valueList = payload;
			Log.log('valueList in socketNotificationReceived: ', this.valueList);
			this.updateDom();
			break
	}
},

/**
 * Prepare the output of the given attribute. Reads the attributeName from the
 * settingsArray and do further processing on it, i.e. to display the value with the
 * unit (temperature, valve state) or anything else.
 * Can be used in the future to prepare any other attributes for output.
 */
prepareAttribute: function(attributeName, strValue, precision, unit){
	var preparedAttributeValue = "";
	switch(attributeName){
		case "TEMPERATURE":
		case "PRESSURE":
		case "HUMIDITY":
			preparedAttributeValue = Number(parseFloat(strValue)).toLocaleString(this.config.localeStr, {minimumFractionDigits: precision, maximumFractionDigits: precision}) + unit;
			break;
	}
	return preparedAttributeValue;
}, 

})

node_helper-async-map.js
/* ---------------------------------------------------------------------
 * Magic Mirror
 * Module: MMM-CaravanPiClimate
 *
 * CaravanPi Module
 * see https://github.com/spitzlbergerj/CaravanPi for more Information 
 *		 about the DIY project 
 *
 * By Josef Spitzlberger	http://spitzlberger.de
 * MIT Licensed.
 */

const NodeHelper = require("node_helper")
var async = require('async');
var exec = require('child_process').exec;
var valueListNH = [];

module.exports = NodeHelper.create({
	start: function() {
		//console.error('Starting node helper: ' + this.name);
	},
	
	socketNotificationReceived: function(notification, payload) {
		var self = this;
		console.error('node_helper: ' + notification);
		
		switch(notification) {
			case "CONFIG":
				this.config = payload.config;
				console.error('node_helper - config: ' + this.config.valueDir);
				
				this.valueListNH = payload.valueList;
				
				// first call
				self.getValues(payload.valueList);
				// interval call
				setInterval(function() {
					self.getValues(payload.valueList);
				}, this.config.updateInterval);
				break
		}
	},
	
	getValues: function(valueList) {
		var self = this;

		console.error('node_helper - getValues - valueList', valueList[0], valueList[1]);

		async.map(valueList, tailValueFile, fillValueList);

		console.error('node_helper - getValues - valueList after', this.valueListNH[0], this.valueListNH[1]);
		self.sendSocketNotification('VALUES', this.valueListNH);
	},
	
	tailValueFile: function(sensor, callback) {
		var self = this;
		var cmd = "tail -1 " + self.config.valueDir + "/" + sensor["file"];
		console.error('node_helper - cmd', cmd);
		
		return exec(cmd, (error, stdout, stderr) => {
			if (error) {
				console.error('exec error: ${error}');
				callback(1, null);
				return;
			}
			console.error('stdout: ${stdout}');
			console.error('stderr: ${stderr}');
			return stdout;
			});
	},
	
	fillValueList: function (err, res) {
		console.error('node_helper - getValues - Results', res);
		var resSplit = res[0].split(' ');
		console.error('node_helper - getValues - Results', resSplit);
		this.valueListNH[i]["datetime"] = resSplit[0].substring(6,8)+"."+resSplit[0].substring(4,6)+"."+resSplit[0].substring(0,4)+" "+resSplit[0].substring(8,10)+":"+resSplit[0].substring(10,13);
		this.valueListNH[i]["temperature"] = resSplit[1];
		this.valueListNH[i]["pressure"] = resSplit[2];
		this.valueListNH[i]["humidity"] = resSplit[3];
	},

})

MMM-CaravanPiTemperature.js
/* ---------------------------------------------------------------------
 * Magic Mirror
 * Module: MMM-CaravanPiTemperature
 *
 * CaravanPi Module
 * see https://github.com/spitzlbergerj/CaravanPi for more Information 
 *     about the DIY project 
 *
 * By Josef Spitzlberger http://spitzlberger.de
 * MIT Licensed.
 */

Module.register("MMM-CaravanPiTemperature",{

defaults:{
	valueDir: "/home/pi/CaravanPi/values",
	updateInterval: 100000, // milliseconds
	tempUnit: " °C",
	tempPrecision: 2,
	showDate: true,
	sensors: [
		{
			name: "Gefrierfach",
			file: "28-01144febdbaa",
		},
	],
	localeStr: 'de-DE',
	style: "lines",
},

valueList:[],

start: function (){
	Log.log('Starting module: ' + this.name);
	
	this.valueList = new Array();
	var i = 0;
	Log.log('sensors: ', this.config.sensors.length, this.config.sensors);
	while(i<this.config.sensors.length){
		this.valueList[i] = new Object();
		this.valueList[i]["name"] = this.config.sensors[i]["name"];
		this.valueList[i]["file"] = this.config.sensors[i]["file"];
		this.valueList[i]["datetime"] = this.translate('LOADING');
		this.valueList[i]["temperature"] = "0";
		i+=1;
	}
	Log.log('valueList: ', this.valueList);
	this.sendSocketNotification(
		'CONFIG',
		{
			config: this.config,
			valueList: this.valueList,
		});
},

// Get translations
getTranslations: function() {
	return {
		en: "translations/en.json",
		de: "translations/de.json"
	}
},

// Get the Module CSS
getStyles: function() {
	return ["MMM-CaravanPiTemperature.css"];
},


getDom: function(){
	var table = document.createElement("table");
	table.border = 0;
	
	if (this.config.style == "boxes") {
		var boxRow = document.createElement("tr");
		boxRow.className = 'sensorBoxRow';
		boxRow.vAlign = 'top';
	}

	var i = 0;
	while (i<this.config.sensors.length) {
		var temperatureStr = this.prepareAttribute("TEMPERATURE", this.valueList[i]["temperature"], this.config.tempPrecision, this.config.tempUnit);

		if (this.config.style == "lines") {
			var row = document.createElement("tr");
			row.className = 'sensorContainer';
			row.vAlign = 'top';
			
			var rowSensor = document.createElement("td");
			rowSensor.className = 'sensorName';
			rowSensor.width = '120px';
			rowSensor.appendChild(document.createTextNode(this.valueList[i]["name"]));
			
			var rowTemperature = document.createElement("td");
			rowTemperature.className = 'sensorTemp';
			rowTemperature.width = '60px';
			rowTemperature.appendChild(document.createTextNode(temperatureStr));
			
			var rowDate = document.createElement("td");
			rowDate.className = 'sensorDate';
			rowDate.width = '60px';
			rowDate.appendChild(document.createTextNode(this.valueList[i]["datetime"]));
			
			// Building of the table row
			row.appendChild(rowSensor);
			row.appendChild(rowTemperature);
			
			if(this.config.showDate === true) {
				row.appendChild(rowDate);
			}
			
			table.appendChild(row);
		}
		else if (this.config.style == "boxes") {
			var boxRowElement = document.createElement("td");
			boxRowElement.style.padding = '20px';
			
			var tableInner = document.createElement("table");
			tableInner.style.border= '1px solid #ffffff';
	
			var row1 = document.createElement("tr");
			row1.className = 'sensorContainer';
			row1.align = 'center';
			row1.vAlign = 'top';
			
			var rowSensor = document.createElement("td");
			rowSensor.className = 'sensorName';
			rowSensor.style.borderBottom = '1px dotted #ffffff';
			rowSensor.appendChild(document.createTextNode(this.valueList[i]["name"]));
			
			row1.appendChild(rowSensor);
			
			var row2 = document.createElement("tr");
			row2.className = 'sensorContainer';
			row2.align = 'center';
			row2.vAlign = 'top';
			
			var rowTemperature = document.createElement("td");
			rowTemperature.className = 'sensorTemp';
			rowTemperature.appendChild(document.createTextNode(temperatureStr));
			
			row2.appendChild(rowTemperature);
			
			var row3 = document.createElement("tr");
			row3.className = 'sensorContainer';
			row3.align = 'center';
			row3.vAlign = 'top';
			
			var rowDate = document.createElement("td");
			rowDate.className = 'sensorDate';
			rowDate.appendChild(document.createTextNode(this.valueList[i]["datetime"]));
			
			row3.appendChild(rowDate);
			
			// Building of the table rows
			tableInner.appendChild(row1);
			tableInner.appendChild(row2);
			
			if(this.config.showDate === true) {
				tableInner.appendChild(row3);
			}
			boxRowElement.appendChild(tableInner)
			boxRow.appendChild(boxRowElement);
		}
		i+=1;
	}
	
	if (this.config.style == "boxes") {
		table.appendChild(boxRow);
	}

	var wrapper = document.createElement("div")
	wrapper.className = "MMM-CaravanPiTemperature";
	
	wrapper.innerHTML = table.outerHTML;
	return wrapper
},

/*
notificationReceived: function(notification, payload, sender){
	switch(notification) {
		case "DOM_OBJECTS_CREATED":
			var timer = setInterval(()=>{
				var countElm = document.getElementById("COUNT")
				this.sendSocketNotification("DO_YOUR_JOB", this.count)
				this.count++
			}, 1000)
			break
	}
},

*/

socketNotificationReceived: function(notification, payload){
	Log.log('MMM-Systemvalues: socketNotificationReceived ' + notification + payload);
	switch(notification) {
		case "VALUES":
			this.valueList = payload;
			Log.log('valueList in socketNotificationReceived: ', this.valueList);
			this.updateDom();
			break
	}
},

/**
 * Prepare the output of the given attribute. Reads the attributeName from the
 * settingsArray and do further processing on it, i.e. to display the value with the
 * unit (temperature, valve state) or anything else.
 * Can be used in the future to prepare any other attributes for output.
 */
prepareAttribute: function(attributeName, strValue, precision, unit){
	var preparedAttributeValue = "";
	switch(attributeName){
		case "TEMPERATURE":
		case "PRESSURE":
		case "HUMIDITY":
			preparedAttributeValue = Number(parseFloat(strValue)).toLocaleString(this.config.localeStr, {minimumFractionDigits: precision, maximumFractionDigits: precision}) + unit;
			break;
	}
	return preparedAttributeValue;
}, 

})